___

<a href='https://mainacademy.ua/'> <img src='https://mainacademy.ua/wp-content/uploads/2019/03/logo-main.png' alt = 'Хмм, щось з інтернетом'/></a>
___

# Module 4: Work with data professionally!

## Lab work 4



#### Мета: 

* навчитися обробляти дані
* "парсити" сайти

### Завдання 1:

#### Ознайомитися із API Національного банку 
https://old.bank.gov.ua/control/uk/publish/article?art_id=38441973&cat_id=38459171#exchange



1. Підключитися до API.


In [7]:
import requests
baneling='https://bank.gov.ua/NBUStatService/v1/statdirectory/exchangenew'
requests.get(baneling)

<Response [200]>

In [8]:
#?valcode=EUR&date=YYYYMMDD

In [9]:
#?date=YYYYMMDD

2. Отримати курс долара за останній рік.

In [10]:
#!pip install datetime
import requests
from tqdm.notebook import tqdm_notebook
import time
import datetime
from calendar import monthrange
def date_rate():
    '''Намутил курсы в словарь от start до end'''
    print('Дані буде завантажено місяцями з start до end поденно')
    baneling='https://bank.gov.ua/NBUStatService/v1/statdirectory/exchangenew'
    start=input('Напиши дату "старт" в форматі YYYYMM (напр.:202001): ')
    end=input('Напиши дату "кінець" в форматі YYYYMM (напр.:202012): ')
    if start =="" and end == "":
        now_y=datetime.datetime.now().year
        start=datetime.date(year=now_y,month=1,day=1)
        end=datetime.datetime.now().date()
        ns=start
        ne=end
    elif start =="" and end !="":
        now_y=datetime.datetime.now().year
        start=datetime.date(year=now_y,month=1,day=1)
        end=end+'05'
        end=datetime.datetime.strptime(end,'%Y%m%d').date()
        end=datetime.date(year=end.year,month=end.month,day=monthrange(end.year, end.month)[-1])
        ns=start
        ne=end
    elif start !="" and end =="":
        start=start+'01'
        start=datetime.datetime.strptime(start,'%Y%m%d').date()
        end=datetime.datetime.now().date()
        ns=start
        ne=end
    else:
        start=start+'01'
        start=datetime.datetime.strptime(start,'%Y%m%d').date()
        end=end+'05'
        end=datetime.datetime.strptime(end,'%Y%m%d').date()
        end=datetime.date(end.year, end.month, monthrange(end.year, end.month)[-1])
        ns=start
        ne=end
    dates=[]
    rate=list()
    params={'date':'Zorro','valcode':'USD','json':'json'}
    while start<=end:
        dates.append(start.strftime('%Y%m%d'))
        start += datetime.timedelta(days=1)
    else:
        counter=int(str(ne-ns).split()[0])+1
        print(f'Буде завантажено дані за {counter} днів, зачекай!')
    for dat in tqdm_notebook(dates,desc='Downloading data strings'):
        params['date']=dat
        rate.append(requests.get(baneling,params=params).json()[0]['rate'])
        time.sleep(0.00001)
    d_r={}
    d_r=dict(zip(dates,rate))
    return d_r



d_r=date_rate()
#d_r




Дані буде завантажено місяцями з start до end поденно
Напиши дату "старт" в форматі YYYYMM (напр.:202001): 202001
Напиши дату "кінець" в форматі YYYYMM (напр.:202012): 202012
Буде завантажено дані за 366 днів, зачекай!


3. Вивести середнє значення та відхилення курсу за кожний місяць.

In [5]:
import datetime
from datetime import datetime
from calendar import monthrange


month_list=[datetime.strptime(list(d_r.keys())[i],'%Y%m%d').strftime('%m') for i in range(0,len(list(d_r.keys())))]

def month_averages():
    solution={}
    global month_average
    month_average={}
    month_change={}
    for mo in set(month_list):
#        print(mo)
        month_average[mo]=[]
        tempo=[]
        for k in d_r.keys():
            if datetime.strptime(k,'%Y%m%d').strftime('%m')==mo:  
                tempo.append(d_r[k])
                month_average[mo]=tempo
                month_average[mo]=round(sum(tempo)/len(tempo),2)
    
    for k,v in d_r.items():
        start=k[:6]+'01'
        end=datetime.strptime(k,'%Y%m%d')
        end=datetime(end.year, end.month, monthrange(end.year, end.month)[-1]).strftime('%Y%m%d')
        if datetime.strptime(end,'%Y%m%d')<datetime.now():
            delta=round((d_r[end]/d_r[start] -1)*100,2)
            month_change[datetime.strptime(end,'%Y%m%d').strftime('%m')]=delta
        else:
            delta=0.00
            month_change[datetime.strptime(end,'%Y%m%d').strftime('%m')]=delta
                   
    for k,v in dict(month_average).items():
        solution[k]=f'середній курс за місяць для USD/UAH {v}, зміна з початку місяця : {month_change[k]}%'
        solution=dict(sorted(solution.items()))
    return solution  
month_averages()

{'01': 'середній курс за місяць для USD/UAH 24.12, зміна з початку місяця : 5.21%',
 '02': 'середній курс за місяць для USD/UAH 24.6, зміна з початку місяця : -1.44%',
 '03': 'середній курс за місяць для USD/UAH 26.41, зміна з початку місяця : 14.25%',
 '04': 'середній курс за місяць для USD/UAH 27.22, зміна з початку місяця : -2.38%',
 '05': 'середній курс за місяць для USD/UAH 26.81, зміна з початку місяця : -0.24%',
 '06': 'середній курс за місяць для USD/UAH 26.71, зміна з початку місяця : -0.66%',
 '07': 'середній курс за місяць для USD/UAH 27.31, зміна з початку місяця : 3.84%',
 '08': 'середній курс за місяць для USD/UAH 27.52, зміна з початку місяця : -0.78%',
 '09': 'середній курс за місяць для USD/UAH 27.98, зміна з початку місяця : 2.67%',
 '10': 'середній курс за місяць для USD/UAH 28.32, зміна з початку місяця : 0.45%',
 '11': 'середній курс за місяць для USD/UAH 28.31, зміна з початку місяця : 0.1%',
 '12': 'середній курс за місяць для USD/UAH 28.17, зміна з початку місяц

4. Дану інформацію записати у файл за допомогою pickle.

In [6]:
import pickle
pickled=month_averages()
with open('pik.pickle', 'wb') as f:
    pickle.dump(pickled, f)


In [7]:
#with open('pik.pickle', 'rb') as f:
#    loaded_pik = pickle.load(f)
#loaded_pik

### Завдання 2:

Потрібно проаналізувати всі товари на сайті: 
https://smallpacking.agrosem.ua/shop/

Створіть базу даних, що містить таблицю із такими стовбцями: 
    - артикул товару
    - найменування
    - ціна 
    - вага

In [8]:
#!pip install bs4
#!pip install lxml

In [9]:
import requests
from bs4 import BeautifulSoup

params={'orderby':'price-desc'}
link = 'https://smallpacking.agrosem.ua/shop/page/2/'
links=[]
i=1
'''Отримую pagination'''
while str(requests.get(link,params=params))== '<Response [200]>':
    
    link='https://smallpacking.agrosem.ua/shop/page/'+str(i)+'/'
    if str(requests.get(link,params=params))== '<Response [200]>':
        links.append(link)
    i+=1  

In [10]:
'''Отримую product link'''
product_link=[]
for i in links:
    request=requests.get(i,params=params)
    s=BeautifulSoup(request.text,'lxml') 
    for b in s.find_all('div',class_='product-from-category-container'):
        product_link.append(b.find('a').get('href'))

In [11]:
'''Парсинг'''
data={}
name_list=[]
SKU_list=[]
price_list=[]
weight_list=[]
for i in tqdm_notebook(product_link,desc='Downloading data strings'):
    request=requests.get(i,params=params)
    s=BeautifulSoup(request.text,'lxml')
    product=s.find('li',class_="active").text
    
    SKU=s.find('span',class_="sku").text
    try:
        price=s.find('span',class_="regular-price").text
    except AttributeError:
        price=None
        pass
    try:
        weight=s.find('tr',class_="woocommerce-product-attributes-item woocommerce-product-attributes-item--attribute_pa_vaga").find('td',class_="woocommerce-product-attributes-item__value").find('p').text
    except AttributeError:
        weight=None
        pass
    name_list.append(product)
    SKU_list.append(SKU)
    price_list.append(price)
    weight_list.append(weight)
    time.sleep(0.00001)
        
data={k:[a,b,c] for k,a,b,c in zip(SKU_list,name_list,price_list,weight_list)}
    

In [12]:
data

{'SKU: 42579': ['Етерно КВС, насіння озимого жита, протруєне.',
  '56 000 грн',
  None],
 'SKU: 12789': ['КВС Пропауер, насіння озимого жита.', '55 000 грн', None],
 'SKU: 62997': ['КВС Тайо, насіння озимого жита.', '55 000 грн', None],
 'SKU: 91815': ['Етерно КВС, насіння озимого жита.', '53 000 грн', None],
 'SKU: 31292': ['Етерно КВС, насіння озимого жита.', '50 000 грн', None],
 'SKU: 56714': ['КВС Боно, насіння озимого жита.', '49 000 грн', None],
 'SKU: 61723': ['КВС Магніфіко, насіння озимого жита на силос.',
  '46 000 грн',
  None],
 'SKU: 35951': ['YaraIrix.', '18 000 грн', '1 шт'],
 'SKU: 63243': ['YaraMila NPK 16-27-7.', '16 800 грн', '600 кг'],
 'SKU: 24344': ['YaraMila NPK 7-20-28.', '16 000 грн', '600 кг'],
 'SKU: 89147': ['YaraMila NPK 12-24-12.', '15 800 грн', '600 кг'],
 'SKU: 64776': ['YaraMila NPK 8-24-24.', '15 600 грн', '600 кг'],
 'SKU: 66872': ['YaraMila NPK 9-12-25 + 2MgO + 6,5SO3 + micro.',
  '13 800 грн',
  '600 кг'],
 'SKU: 84721': ['YaraMila NPK 16-16-16.', 

### Завдання 3: 

Ознайомтеся із роботою SQLite та відповідним модулем у Python.
Завантажте базу даних для виконання лабораторної роботи.
Підключіться до завантаженої бази SQLite.

1. Виведіть інформацію про дану базу.

In [10]:
import sqlite3#1.import

sqlite_connection = sqlite3.connect('tysql.sqlite')#2.instance and connect bd name
cursor = sqlite_connection.cursor()

#sqlite_select_query = "select sqlite_version();"
sqlite_select_query="SELECT * FROM sqlite_master"
cursor.execute(sqlite_select_query)#3 сделать запрос и получить resultSet
record = cursor.fetchall()#4 .fetchall() получить из resultSet результат запроса
for row in record:
    print (row[1])
    print (row[2])
    print (row[4])
cursor.close()


Customers
Customers
CREATE TABLE Customers
(
  cust_id      char(10)  NOT NULL ,
  cust_name    char(50)  NOT NULL ,
  cust_address char(50)  NULL ,
  cust_city    char(50)  NULL ,
  cust_state   char(5)   NULL ,
  cust_zip     char(10)  NULL ,
  cust_country char(50)  NULL ,
  cust_contact char(50)  NULL ,
  cust_email   char(255) NULL ,
  PRIMARY KEY (cust_id)
)
sqlite_autoindex_Customers_1
Customers
None
OrderItems
OrderItems
CREATE TABLE OrderItems
(
  order_num  int          NOT NULL                      ,
  order_item int          NOT NULL                      ,
  prod_id    char(10)     NOT NULL                      ,
  quantity   int          NOT NULL                      ,
  item_price decimal(8,2) NOT NULL                      ,
  PRIMARY KEY (order_num, order_item)                   ,
  FOREIGN KEY (order_num) REFERENCES Orders (order_num) ,
  FOREIGN KEY (prod_id) REFERENCES Products (prod_id)
)
sqlite_autoindex_OrderItems_1
OrderItems
None
Orders
Orders
CREATE TABLE Orders

2. Виведіть перелік всіх таблиць.

In [11]:
sqlite_connection = sqlite3.connect('tysql.sqlite')#2.instance and connect bd name
cursor = sqlite_connection.cursor()
cursor.execute("select * from sqlite_master where type = 'table'")
tables = cursor.fetchall()

for table in tables:
    print(table[1]) #названия тадлиц

Customers
OrderItems
Orders
Products
Vendors


3. Список всіх cust_id з таблиці Customers table.

In [12]:
cursor.execute('SELECT cust_id FROM Customers')
tables = cursor.fetchall()
tables

[('1000000001',),
 ('1000000002',),
 ('1000000003',),
 ('1000000004',),
 ('1000000005',)]

4. Всю таблицю Customers table

In [13]:
cursor.execute('SELECT * FROM Customers')
tables = cursor.fetchall()
tables

[('1000000001',
  'Village Toys',
  '200 Maple Lane',
  'Detroit',
  'MI',
  '44444',
  'USA',
  'John Smith',
  'sales@villagetoys.com'),
 ('1000000002',
  'Kids Place',
  '333 South Lake Drive',
  'Columbus',
  'OH',
  '43333',
  'USA',
  'Michelle Green',
  None),
 ('1000000003',
  'Fun4All',
  '1 Sunny Place',
  'Muncie',
  'IN',
  '42222',
  'USA',
  'Jim Jones',
  'jjones@fun4all.com'),
 ('1000000004',
  'Fun4All',
  '829 Riverside Drive',
  'Phoenix',
  'AZ',
  '88888',
  'USA',
  'Denise L. Stephens',
  'dstephens@fun4all.com'),
 ('1000000005',
  'The Toy Store',
  '4545 53rd Street',
  'Chicago',
  'IL',
  '54545',
  'USA',
  'Kim Howard',
  None)]

5. Список клієнтів (cust_names) відсортованих від Z до A

In [14]:
cursor.execute('SELECT cust_name FROM Customers ORDER BY cust_name DESC')
tables = cursor.fetchall()
tables

[('Village Toys',),
 ('The Toy Store',),
 ('Kids Place',),
 ('Fun4All',),
 ('Fun4All',)]

6. Таблицю клієнтів та замовлень (cust_id і order_num) . Відсортуйте по клієнту і потім по даті замовлення

In [24]:
cursor.execute("""
SELECT Customers.cust_id, Customers.cust_name, Orders.order_num, Orders.order_date
FROM  Customers LEFT JOIN Orders ON Customers.cust_id = Orders.cust_id 
                            ORDER BY Customers.cust_id,Orders.order_num
               
               """)

tables = cursor.fetchall()
print('(cust_id і order_num) sorted:')
tables

(cust_id і order_num) sorted:


[('1000000001', 'Village Toys', 20005, '2012-05-01'),
 ('1000000001', 'Village Toys', 20009, '2012-02-08'),
 ('1000000002', 'Kids Place', None, None),
 ('1000000003', 'Fun4All', 20006, '2012-01-12'),
 ('1000000004', 'Fun4All', 20007, '2012-01-30'),
 ('1000000005', 'The Toy Store', 20008, '2012-02-03')]

сортирован по cust_id
[('1000000001', 'Village Toys', 20005, '2012-05-01'),
 ('1000000001', 'Village Toys', 20009, '2012-02-08'),
 ('1000000002', 'Kids Place', None, None),
 ('1000000003', 'Fun4All', 20006, '2012-01-12'),
 ('1000000004', 'Fun4All', 20007, '2012-01-30'),
 ('1000000005', 'The Toy Store', 20008, '2012-02-03')]

сортирован по order_num
[('1000000002', 'Kids Place', None, None),
 ('1000000001', 'Village Toys', 20005, '2012-05-01'),
 ('1000000003', 'Fun4All', 20006, '2012-01-12'),
 ('1000000004', 'Fun4All', 20007, '2012-01-30'),
 ('1000000005', 'The Toy Store', 20008, '2012-02-03'),
 ('1000000001', 'Village Toys', 20009, '2012-02-08')]

In [ ]:
#Но нужно использовать функцию executemany вместо обычной execute, fetchmany() fetchall()  fetchone()заметка для себя


7. Таблицю (на основі Items) з кількість та вартістю товару. Відсортуйте в порядку спадання по кількості та вартості

In [26]:
cursor.execute("""
SELECT quantity, item_price
FROM OrderItems
ORDER BY quantity desc, item_price desc
               
               """)

tables = cursor.fetchall()
print('( quantity desc, item_price) sorted:')
tables

( quantity desc, item_price) sorted:


[(250, 2.49),
 (250, 2.49),
 (250, 2.49),
 (100, 10.99),
 (100, 5.49),
 (100, 2.99),
 (100, 2.99),
 (100, 2.99),
 (50, 11.49),
 (50, 4.49),
 (20, 5.99),
 (10, 11.99),
 (10, 8.99),
 (10, 3.49),
 (10, 3.49),
 (10, 3.49),
 (5, 11.99),
 (5, 4.99)]

8. Товар (з таблиці Products), ціна якого становить 9.49

In [28]:
cursor.execute("""
SELECT prod_id
FROM Products
WHERE prod_price = 9.49
               
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[('RYL01',), ('RYL02',)]

9. Виведіть назву товару та ціну, яка лежить в діапазоні від  3 до 6. Відсортуйте результат в по ціні в порядку зростання

In [32]:
cursor.execute("""
SELECT prod_id, prod_price
FROM Products
WHERE prod_price >= 3 and prod_price <=6
ORDER BY prod_price ASC               
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

( quantity desc, item_price) sorted:


[('BNBG01', 3.49),
 ('BNBG02', 3.49),
 ('BNBG03', 3.49),
 ('RGAN01', 4.99),
 ('BR01', 5.99)]

10. Кількість товару, що було продано

In [35]:
cursor.execute("""
SELECT sum(quantity)
FROM OrderItems
          
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[(1430,)]

11. Кількість найменувань товару, ціна якого більша за 4

In [36]:
cursor.execute("""
SELECT count(prod_id)
FROM OrderItems
WHERE item_price >4
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[(9,)]

12. Розробіть алгоритм для виведення 3 найдорожчих товарів в базі

In [43]:
cursor.execute("""
SELECT prod_id, prod_price 
FROM Products
ORDER BY prod_price DESC
LIMIT 3
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[('BR03', 11.99), ('RYL01', 9.49), ('RYL02', 9.49)]

13. Підрахуйте кількість замовлень для кожного клієнта використовуючи підзапити

In [46]:
cursor.execute("""
SELECT Customers.cust_id,Customers.cust_name,(
    SELECT count(Orders.order_num)
    FROM Orders
    WHERE Customers.cust_id = Orders.cust_id)
FROM Customers
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[('1000000001', 'Village Toys', 2),
 ('1000000002', 'Kids Place', 0),
 ('1000000003', 'Fun4All', 1),
 ('1000000004', 'Fun4All', 1),
 ('1000000005', 'The Toy Store', 1)]

14. Виведіть список клієнтів та їх замовлення (використайте JOIN)

In [48]:
cursor.execute("""
SELECT Customers.cust_id,Customers.cust_name, Orders.order_num
FROM Customers 
LEFT JOIN Orders ON Customers.cust_id = Orders.cust_id
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[('1000000001', 'Village Toys', 20005),
 ('1000000001', 'Village Toys', 20009),
 ('1000000002', 'Kids Place', None),
 ('1000000003', 'Fun4All', 20006),
 ('1000000004', 'Fun4All', 20007),
 ('1000000005', 'The Toy Store', 20008)]

15. Зробіть два запити: список товарів, ціна яких менша 5 та список товарів, ціна який більша рівна 5. Використайте UNION для об'єднання цих запитів

In [56]:
cursor.execute("""
SELECT prod_name, prod_price
FROM Products
WHERE prod_price < 5
UNION
SELECT prod_name, prod_price
FROM Products
WHERE prod_price >= 5 
               """)

tables = cursor.fetchall()
#print('( quantity desc, item_price) sorted:')
tables

[('Fish bean bag toy', 3.49),
 ('Bird bean bag toy', 3.49),
 ('Rabbit bean bag toy', 3.49),
 ('Raggedy Ann', 4.99),
 ('8 inch teddy bear', 5.99),
 ('12 inch teddy bear', 8.99),
 ('18 inch teddy bear', 11.99),
 ('King doll', 9.49),
 ('Queen doll', 9.49)]

### Завдання 4: 

Зареєструйтеся на сайті  http://www.sql-ex.ru/

Виконайте завдання із блоку SELECT; з 1 по 30 завдання

Код запитів вставляйте сюди


In [ ]:
'''SELECT
PC.model, PC.speed, PC.hd
FROM PC
WHERE price < 500
SELECT DISTINCT Product.maker

FROM Product
WHERE Product.type = 'Printer'
SELECT Laptop.model, Laptop.ram, Laptop.screen
FROM Laptop
WHERE Laptop.price > 1000
SELECT *
FROM Printer
WHERE color = 'y'
SELECT DISTINCT model, speed, hd
FROM PC
WHERE (cd = '12x' or cd = '24x') and price <600
SELECT DISTINCT Product.maker,  Laptop.speed
FROM Product
LEFT JOIN Laptop on Laptop.model =  Product.model
WHERE Laptop.hd >= 10
SELECT DISTINCT product.model, pc.price
FROM Product JOIN pc ON product.model = pc.model WHERE maker = 'B'
UNION
SELECT DISTINCT product.model, laptop.price
FROM product JOIN laptop ON product.model=laptop.model WHERE maker='B'
UNION
SELECT DISTINCT product.model, printer.price
FROM product JOIN printer ON product.model=printer.model WHERE maker='B'
SELECT DISTINCT Product.maker
FROM Product LEFT JOIN PC on PC.model=Product.model
WHERE PC.speed>=450
SELECT model, price
FROM Printer
WHERE price = (SELECT MAX(price) FROM Printer)
SELECT AVG(speed)
FROM pc
SELECT AVG(Laptop.speed)
FROM Laptop
WHERE price > 1000
SELECT AVG(PC.speed)
FROM PC LEFT JOIN Product ON Product.model=PC.model
WHERE maker = 'A'

SELECT Ships.class,Ships.name,Classes.country
FROM Ships LEFT JOIN Classes on Ships.class = Classes.class
WHERE Classes.numGuns >= 10
SELECT DISTINCT HD
FROM PC
GROUP BY HD
HAVING count(HD) >=2
SELECT DISTINCT Product.type,Laptop.model,Laptop.speed
FROM Laptop, PC, Product
WHERE type = 'Laptop' AND Laptop.speed < ALL (SELECT PC.speed FROM PC)
SELECT DISTINCT maker, price
FROM ( SELECT Product.maker, Printer.price, Product.type
FROM Product LEFT JOIN Printer on Product.model = Printer.model
WHERE Product.type = 'Printer' and Printer.color = 'y' and Printer.price = (SELECT MIN(price) FROM Printer WHERE Printer.color = 'y')
) AS Winter
SELECT
product.maker, AVG(screen)
FROM laptop
LEFT JOIN product ON product.model = laptop.model
GROUP BY product.maker
SELECT maker, COUNT(model)
FROM product
WHERE type = 'pc'
GROUP BY product.maker
HAVING COUNT(model) >= 3
SELECT product.maker, MAX(pc.price)
FROM product, pc
WHERE product.model = pc.model
GROUP BY product.maker
SELECT speed, AVG(price)
FROM PC
WHERE speed >600
GROUP BY speed
SELECT DISTINCT maker
FROM product t1 JOIN pc t2 ON t1.model=t2.model
WHERE speed>=750 AND maker IN
( SELECT maker
FROM product t1 JOIN laptop t2 ON t1.model=t2.model
WHERE speed>=750 )
SELECT DISTINCT model
FROM (
	SELECT model, price
	FROM PC
	UNION ALL
	SELECT model, price
	FROM Laptop
	UNION ALL
	SELECT model, price
	FROM Printer
	)X
WHERE price = (SELECT MAX(price) FROM (
	SELECT model, price
	FROM PC
	UNION ALL
	SELECT model, price
	FROM Laptop
	UNION ALL
	SELECT model, price
	FROM Printer
	)X)
SELECT DISTINCT maker
FROM product
WHERE model IN (
SELECT model
FROM pc
WHERE ram = (
  SELECT MIN(ram)
  FROM pc
  )
AND speed = (
  SELECT MAX(speed)
  FROM pc
  WHERE ram = (
   SELECT MIN(ram)
   FROM pc
   )
  )
)
AND
maker IN (
SELECT maker
FROM product
WHERE type='printer'
)
SELECT product.maker, AVG(pc.hd)
FROM pc, product WHERE product.model = pc.model
AND product.maker IN ( SELECT DISTINCT maker
FROM product
WHERE product.type = 'printer')
GROUP BY maker
SELECT COUNT(maker)
FROM (
SELECT DISTINCT maker, COUNT(model) qty
FROM Product
GROUP BY maker
HAVING COUNT(model) = 1)X
SELECT t1.point, t1.date, inc, out
FROM income_o t1 LEFT JOIN outcome_o t2 ON t1.point = t2.point
AND t1.date = t2.date
UNION
SELECT t2.point, t2.date, inc, out
FROM income_o t1 RIGHT JOIN outcome_o t2 ON t1.point = t2.point
AND t1.date = t2.date

select point, date, SUM(sum_out), SUM(sum_inc)
from( select point, date, SUM(inc) as sum_inc, null as sum_out from Income Group by point, date
Union
select point, date, null as sum_inc, SUM(out) as sum_out from Outcome Group by point, date ) as t
group by point, date order by point
'''